### PR TITLE
[011] Json File Source

### PR DESCRIPTION
**Pull request by gurbuzali**
add jackson-jr-all as a shaded library
add option to create FileSource for json files with object mapping

Add description here

Checklist
- [X] Tags Set
- [X] Milestone Set
- [x] Any breaking changes are documented
- [x] New public APIs have `@Nonnull/@Nullable` annotations
- [X] New public APIs have `@since` tags in Javadoc
- [x] For code samples, code sample main readme is updated

Links to issues fixed (if any):

Fixes #NNNN

List of breaking changes:

* ..

### CODE REVIEW

**Comment by gurbuzali** [comment]
verify

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/FileSourceBuilder.java [hunk 1]
```diff
@@ -161,6 +168,31 @@ public FileSourceBuilder charset(@Nonnull Charset charset) {
...
+     * <p>
+     * The default local parallelism for this processor is 2 (or 1 if just 1
+     * CPU is available).
+     *
+     */
+    public <T> BatchSource<T> buildJson(@Nonnull Class<T> type) {
+        String charsetName = charset.name();
+        return build(path -> {
+            InputStreamReader reader = new InputStreamReader(new FileInputStream(path.toFile()), charsetName);
+            ValueIterator<T> valueIterator = JSON.std.beanSequenceFrom(type, reader);
```
Comment (jbartok) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/FileSourceBuilder.java:189]: Maybe a JSON manipulation should be inside the `JsonUtil` class and Jackson jr not be exposed anywhere else?

File: hazelcast-jet-core/src/test/java/com/hazelcast/jet/impl/connector/ReadFilesPTest.java [hunk 2]
```diff
@@ -121,10 +148,78 @@ private Pipeline buildDag(String glob) {
...
+    }
+
+    private void finishDirectory(File... files) {
         for (File file : files) {
             assertTrue(file.delete());
         }
         assertTrue(directory.delete());
     }
+
+    public static class TestPerson implements Serializable {
```
Comment (jbartok) [inline hazelcast-jet-core/src/test/java/com/hazelcast/jet/impl/connector/ReadFilesPTest.java:167]: Maybe use Jackson jr annotation support to get rid of all those getters & setters?

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/json/JsonUtil.java [hunk 3]
```diff
@@ -28,14 +32,67 @@ private JsonUtil() {
...
+     */
+    public static Map<String, Object> parse(String jsonString) {
+        return uncheckCall(() -> JSON.std.with(Feature.READ_ONLY).mapFrom(jsonString));
+    }
+
+    /**
+     * Extracts a string value from given JSON string.
+     */
+    public static String getString(String jsonString, String key) {
+        return (String) parse(jsonString).get(key);
```
[the remark this answers is not in the document]
Comment (gurbuzali) [reply hazelcast-jet-core/src/main/java/com/hazelcast/jet/json/JsonUtil.java:75]: yes you are right but if a user wants only a single field from a json string these may come handy, saves you a lot of boilerplate code and type casting. Maybe we should mention that each call will parse the json again.

[the remark this answers is not in the document]
Comment (jbartok) [reply hazelcast-jet-core/src/main/java/com/hazelcast/jet/json/JsonUtil.java:75]: Makes sense.

**Comment by cangencer** [comment]
can we assign the numeric id for the design doc?

**Pull request merged**